In Type.FullName, escape ,+&@*[]\ with blackslash. Closes #26384

lambdageek opened the pull request:
- [x] Enable roundtrip between `Type.FullName` and `Type.GetType(String)`.
 - [x] Git rid of ad-hoc parser in `ModuleBuilder.GetType(String)` use `TypeSpec.Parse`
 - [x] Return display (escaped) name for `TypeBuilder.FullName`
 - [x] Added a wrapper `TypeIdentifier` interface that has a `DisplayName` and an `InternalName` property and use that in `TypeSpec`,  `ModuleBuilder` and `TypeBuilder` instead of bare `string`s.
 - [x] ~~Maybe reflect `TypeSpec` into the native side and get rid of the parser in `mono_reflection_parse_type()`~~

    ~~(Though there's a tricky bit because the native parser is also called from `mini/debugger-agent.c` ... maybe it's not worth it)~~ (Seems like a bigger project. Better to approach is separately.)

kumpera commented on mono/metadata/class.c:
```diff
@@ -427,16 +464,23 @@ mono_type_get_name_recurse (MonoType *type, GString *str, gboolean is_recursed,
+			GString *esc_ns = mono_identifier_escape_type_name_chars (klass->name_space);
+			g_string_append (str, esc_ns->str);
 			g_string_append_c (str, '.');
+			g_string_free(esc_ns, TRUE);
```
Pass the GString 'str' to mono_identifier_escape_type_name_chars. This avoids the intermediary GString.

kumpera commented on mono/metadata/reflection.c:
```diff
@@ -7627,10 +7627,91 @@ _mono_reflection_parse_type (char *name, char **endptr, gboolean is_recursed,
+void mono_identifier_unescape_info (MonoTypeNameParse* info);
+
+static
+void unescape_each_type_argument(void* data, void* user_data)
```
Our coding convention is for the return type plus all modifiers to be in the first line. So it should be:
static void 
unescape_each_type_argument(void* data, void* user_data)

kumpera commented on mono/metadata/reflection.h:
```diff
@@ -136,6 +136,8 @@ MONO_API MonoType* mono_reflection_type_get_type (MonoReflectionType *reftype);
 
 MONO_API MonoAssembly* mono_reflection_assembly_get_assembly (MonoReflectionAssembly *refassembly);
 
+MONO_API char* mono_identifier_unescape_type_name_chars (char* identifier);
```
Let's not add this to the public API unless we're forced to.

kumpera commented on mcs/class/corlib/System/TypeIdentifier.cs:
```diff
@@ -0,0 +1,234 @@
+				return other != null && DisplayName == other.DisplayName;
+			}
+
+			override public int GetHashCode ()
```
the visibility modifier always comes first.

so it's public override.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -81,35 +79,71 @@ internal class PointerSpec : ModifierSpec
+				if (nested != null)
+					return nested;
+				else
+					return new TypeName[] { };
```
Use EmptyArray<TypeName>.Value to avoid excessive allocations.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -81,35 +79,71 @@ internal class PointerSpec : ModifierSpec
+				if (modifier_spec != null)
+					return modifier_spec;
+				else
+					return new ModifierSpec[] { };
```
Same thing as the previous one, lets avoid the extra allocations here.

Quite a bit of corlib performance surprisingly comes from being very conservative in avoiding those
small, insignificant, allocations. In the aggregate they actually amount to be measurable.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -81,35 +79,71 @@ internal class PointerSpec : ModifierSpec
 			if (nested != null) {
 				foreach (var n in nested)
-					str += "+" + n;
+					str += "+" + n.DisplayName;
```
Use a StringBuilder here instead of doing multiple concats.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -152,11 +193,69 @@ internal static TypeSpec Parse (string typeName)
 
+		internal static string EscapeDisplayName(string internalName)
+		{
+			String res = "";
```
Same thing, use a StringBuilder to avoid all the temps.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -152,11 +193,69 @@ internal static TypeSpec Parse (string typeName)
+
+		internal static string UnescapeInternalName(string displayName)
+		{
+			String res = "";
```
Use StringBuilder here too.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -397,6 +504,45 @@ static TypeSpec Parse (string name, ref int p, bool is_recurse, bool allow_aqn)
+			other.nested = nested;
+			other.generic_params = generic_params;
+
+			other.modifier_spec = null;
```
null and false are default values, no need to set to them right after you called the constructor.

kumpera commented on mcs/class/corlib/System/TypeSpec.cs:
```diff
@@ -397,6 +504,45 @@ static TypeSpec Parse (string name, ref int p, bool is_recurse, bool allow_aqn)
+			return other;
+		}
+
+		internal TypeName TypeName {
```
This is not really a getter as it's constructing a new object on each request.

kumpera commented on mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs:
```diff
@@ -376,44 +369,33 @@ public override Type GetType (string className, bool throwOnError, bool ignoreCa
 
 			if (!ignoreCase) {
-				result =  name_cache [className] as TypeBuilder;
+				var displayNestedName = ts.WithoutModifiers().TypeName;
```
I think this could be accomplished in a much more straightforward way than this setup that will:

1)create a new TypeSpec
2)create a new TypeSpecTypeName
3)use TypeSpecTypeName just for this purpose

How about a GetDisplayFullName variant that ignores modifiers? It would avoid all the intermediate steps.

lambdageek replied:
Yea, there's not really a good payoff to `WithoutModifiers()`.  I thought I'd get to use it in a couple of different places, but there really aren't a lot of other APIs that need to manipulate type names in a very fancy way.  I'll make something simpler just for this one use.

kumpera commented on mcs/class/corlib/corlib.dll.sources:
```diff
@@ -142,6 +142,7 @@ System/TimeZoneInfo.Android.cs
 System/TimeZoneInfo.MonoTouch.cs
 System/TimeZoneInfo.Serialization.cs
 ../../build/common/MonoTODOAttribute.cs
+System/TypeIdentifier.cs
```
probably needs to be added to other profiles.

lambdageek replied:
Does that just mean the other `*.sources` files in this directory? Or are there other places that need to change?

kumpera replied:
The other *.sources files in corlib. I'll explain how that works.

kumpera commented on mono/metadata/class.c:
```diff
@@ -427,16 +471,20 @@ mono_type_get_name_recurse (MonoType *type, GString *str, gboolean is_recursed,
 		if (format == MONO_TYPE_NAME_FORMAT_IL) {
-			char *s = strchr (klass->name, '`');
-			int len = s ? s - klass->name : strlen (klass->name);
+			GString *esc = g_string_new ("");
```
I don't think MONO_TYPE_NAME_FORMAT_IL should have names escaped.

This is used in loader.c::find_method and might actually break loading.

lambdageek commented on mcs/class/corlib/System.Reflection.Emit/ModuleBuilder.cs:
```diff
@@ -376,44 +369,33 @@ public override Type GetType (string className, bool throwOnError, bool ignoreCa
-					return mt;
+				throw new TypeLoadException (className);
+			if (result != null && (ts.HasModifiers || ts.IsByRef)) {
+				Type mt = result.CreateType();
```
Okay, I got this pretty wrong. `create_modified_type()` did not cause the `TypeBuilder` to be created.  It did something pretty weird - it grabbed the (not finished) `MonoType` of the typebuilder (setup by `TypeBuilder.setup_internal_class`) and started creating pointer and array types from that `MonoType` and then turning the result of that process into a `MonoReflectionType` (which gets cached by the monodomain so in the future callers get the same reflection type back).

So you got this weird situation where on the managed side you'd get a perfectly usable `MonoType` for the pointer or the array, but if you called `GetElementType()` on that pointer or array, you'd get back the `Type` that the `TypeBuilder` would have created, except that it's in a weird not-created state.

lambdageek commented:
1. Went back to `create_modified_types()` icall in `ModuleBuilder` for adding modifiers. My managed code version was too eager in calling `CreateType()`.
2. `StringBuilder`s everywhere instead of `String`s
3. Added `TypeIdentifier.cs` to the `*.csproj` files.
4. Cleaned up misc new object allocations.
5. code style cleanups.
